### PR TITLE
c api: Introduce C/C++ specific handling of public enums.

### PR DESCRIPTION
**Pull request by aniemetz**
This introduces a general way to use public enums both in the C++ and the (upcoming) C API. For the C++ case, we now use enum classes rather than enums for public enums. This also includes definitions for C API to_string conversions for public enums. C API definitions in cvc5_types.h are only included from the C API, guarded via a macro (thus, for now, not included yet when the header is included).

### CODE REVIEW

**Comment by ajreynol** [comment]
As I understand it, the C++ API now uses enum classes? If so, we should take this opportunity to remove spurious common prefixes from the names of the enums, in `LearnedLitType`, `ProofComponent`, `FindSynthTarget`.

**Comment by aniemetz** [comment]
> As I understand it, the C++ API now uses enum classes? If so, we should take this opportunity to remove spurious common prefixes from the names of the enums, in `LearnedLitType`, `ProofComponent`, `FindSynthTarget`.

Yes, that's right. And good point. I will make these changes.

File: NEWS.md [hunk 1]
```diff
@@ -10,20 +11,20 @@ This file contains a summary of important user-visible changes.
...
+  * API: New API function `Solver::findSynth` which takes an identifier
     specifying a target term to synthesize and (optionally) a grammar. This
     method can be used to directly enumerate terms in a provided grammar
     (`FIND_SYNTH_TARGET_ENUM`), or as a way of finding other terms of interest,
     e.g. a rewrite rule that is applicable to the current set of assertions
     (`FIND_SYNTH_TARGET_REWRITE_INPUT`).
   - API: New API function `Solver::findSynthNext` which gets the next term
     in the enumeration.
-  - SMT-LIB: New commands `find-synth` and `find-synth-next` which invoke the
+  * SMT-LIB: New commands `find-synth` and `find-synth-next` which invoke the
```
Comment (ajreynol) [inline NEWS.md:27]: Also change the bullet on line 25.

**Review by ajreynol — APPROVED** [review]
Since this is a major change to the user (in particular requiring Kind::), I assume we should prioritize pushing for a 1.1 release soon?

File: include/cvc5/cvc5_types.h [hunk 2]
```diff
@@ -301,11 +448,42 @@ enum FindSynthTarget
...
+
+#ifndef CVC5_API_USE_C_ENUMS
+#ifndef CVC5__API__CVC5_CPP_TYPES_H
+#define CVC5__API__CVC5_CPP_TYPES_H
+#endif
+#else
+#ifndef CVC5__API__CVC5_C_TYPES_H
+#define CVC5__API__CVC5_C_TYPES_H
+#endif
+#endif
```
Comment (mpreiner) [inline include/cvc5/cvc5_types.h:489]:
```suggestion
#ifdef CVC5_API_USE_C_ENUMS
#ifndef CVC5__API__CVC5_C_TYPES_H
#define CVC5__API__CVC5_C_TYPES_H
#endif
#else
#ifndef CVC5__API__CVC5_CPP_TYPES_H
#define CVC5__API__CVC5_CPP_TYPES_H
#endif
#endif
```

File: include/cvc5/cvc5_kind.h [hunk 3]
```diff
@@ -5873,20 +5934,42 @@ std::string sortKindToString(SortKind k) CVC5_EXPORT;
...
+
+#ifndef CVC5_API_USE_C_ENUMS
+#ifndef CVC5__API__CVC5_CPP_KIND_H
+#define CVC5__API__CVC5_CPP_KIND_H
+#endif
+#else
+#ifndef CVC5__API__CVC5_C_KIND_H
+#define CVC5__API__CVC5_C_KIND_H
+#endif
+#endif
```
Comment (mpreiner) [inline include/cvc5/cvc5_kind.h:5975]:
```suggestion
#ifdef CVC5_API_USE_C_ENUMS
#ifndef CVC5__API__CVC5_C_KIND_H
#define CVC5__API__CVC5_C_KIND_H
#endif
#else
#ifndef CVC5__API__CVC5_CPP_KIND_H
#define CVC5__API__CVC5_CPP_KIND_H
#endif
#endif
```

**Comment by mpreiner** [comment]
We should only merge this after we did a patch release.

File: NEWS.md [hunk 4]
```diff
@@ -10,20 +11,20 @@ This file contains a summary of important user-visible changes.
...
     returns a subset of the current assertions that cause the solver to timeout
     without a provided timeout (option `--timeout-core-timeout`).
-  - SMT-LIB: New command `(get-timeout-core)` which invokes the above method.
+  * SMT-LIB: New command `(get-timeout-core)` which invokes the above method.
 - Support for new interfaces to the SyGuS solver.
-  - API: New API function `Solver::findSynth` which takes an identifier
+  * API: New API function `Solver::findSynth` which takes an identifier
     specifying a target term to synthesize and (optionally) a grammar. This
     method can be used to directly enumerate terms in a provided grammar
     (`FIND_SYNTH_TARGET_ENUM`), or as a way of finding other terms of interest,
```
Comment (ajreynol) [inline NEWS.md:22]:
```suggestion
    (`FindSynthTarget::ENUM`), or as a way of finding other terms of interest,
```

File: NEWS.md [hunk 5]
```diff
@@ -10,20 +11,20 @@ This file contains a summary of important user-visible changes.
...
     e.g. a rewrite rule that is applicable to the current set of assertions
     (`FIND_SYNTH_TARGET_REWRITE_INPUT`).
```
Comment (ajreynol) [inline NEWS.md:24]:
```suggestion
    (`FindSynthTarget::REWRITE_INPUT`).
```

**Review by ajreynol — CHANGES_REQUESTED** [review]

**Review by ajreynol — APPROVED** [review]

**Pull request merged**